### PR TITLE
feat(gengapic): use gax.BuildHeaders and gax.InsertMetadataIntoOutgoingContext

### PR DESCRIPTION
**Pull request by quartzmo**
closes: #1300
closes: #1301

### CODE REVIEW

**Review by noahdietz — COMMENTED** [review]
I think you trying to do grpc in one PR and rest in another - i'd just do them both in the same PR

**Comment by quartzmo** [comment]
> I think you trying to do grpc in one PR and rest in another - i'd just do them both in the same PR

I tried, but REST is much more involved. So I opened this PR for gRPC to get feedback. If it looks good, I can add the REST changes, and update the title, etc.

How does this look to you so far?

**Comment by noahdietz** [comment]
Ok let me take a deeper look, I just skimmed it before.

**Comment by quartzmo** [comment]
@noahdietz ping

File: go.sum
Comment (noahdietz) [inline go.sum:1]: I would revert this and catch up with `main` it's creating conflicts atm

File: showcase/go.sum
Comment (noahdietz) [inline showcase/go.sum:1]: I would revert this and catch up with `main` it's creating conflicts atm

**Review by noahdietz — COMMENTED** [review]
looking OK so far, yeah

File: showcase/go.sum
Comment (quartzmo) [reply showcase/go.sum:1]: Done, thanks!

File: go.sum
Comment (quartzmo) [reply go.sum:1]: Done, thanks!

File: internal/gengapic/gengapic.go
Comment (noahdietz) [inline internal/gengapic/gengapic.go:387]: I kind of wish `"Content-Type", "application/json"` was just part of `c.xGoogHeaders` ya know? It's set on every regapic request...but `xGoogHeaders` has always been `x-goog-api-client` headers only...idk wdyt? @codyoss hbu?

**Review by noahdietz — COMMENTED** [review]
LGTM just one question that isn't really blocking

File: internal/gengapic/gengapic.go
Comment (quartzmo) [reply internal/gengapic/gengapic.go:387]: I'll wait to hear back from Cody on this before merging.

Comment (codyoss) [reply internal/gengapic/gengapic.go:387]: I would be fine with it just being agnostic and just general headers.

Comment (tritone) [reply internal/gengapic/gengapic.go:387]: I would imagine `XGoogHeaders` is specifically for headers with the x-goog prefix (of which there are many, not just `x-goog-api-client`)

Comment (quartzmo) [reply internal/gengapic/gengapic.go:387]: @noahdietz What's your call on this?

Comment (noahdietz) [reply internal/gengapic/gengapic.go:387]: Let's just keep it as is.

**Review by noahdietz — APPROVED** [review]

**Pull request merged**
